### PR TITLE
Introduce pre-commit, black, isort, ...

### PR DESCRIPTION
**Pull request by mauwii**
This PR aims to make things easier for new contributors as well as sort out common problems before commiting.

A good strategy how to do the integration of black into an existing repository can be found [here](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html). It explains how to not "burn" the git blame when adding a huge re-formatting commit. Currently if I run a `pre-commit run --all-files` there would get 208 files changed...

### CODE REVIEW

**Comment by JPPhoto** [comment]
Do you mind if I make some edits on the contributors document?

**Comment by mauwii** [comment]
Feel free, no prob. Now that I know I will leave it untouched till you're done.

I was also already thinking about if instead of just having one document, make a Contributors Folder with different toppics, like the already begun "how to create a PR", but also pointing to possible topics for contributions, like the translations, GH-Actions, u name it, since this would maybe help to encourage / enable people who do not have a background in development but still would like to help out 😅

**Comment by JPPhoto** [comment]
I think that's a good evolution. We can refactor the doc at some point if needed! I want to split up what you've written into a non-technical (about the project, philosophy, etc.) and technical and expand where I can.

**Comment by mauwii** [comment]
ping @JPPhoto - will you commit the changes to this draft PR, or can I continue updating `010_HOW_TO_CONTRIBUTE.md`?

**Comment by lstein** [comment]
Is this still being worked on?

**Comment by mauwii** [comment]
@lstein Kind of yes - it will imho not be from value to merge this PR without enabling the pre-commit hooks, but enabling them comes with the price that black and isort would be enabled, which was decided should happen after the refactoring (at least this is the latest information I got regarding this topic)

**Comment by mauwii** [comment]
This PR would also close #2522

**Review by tildebyte — COMMENTED** [review]
LGTM, but I'll leave it to others who actually "have a dog in this fight" to ACK 😁

**Comment by mauwii** [comment]
Nice to finally hear from you, hope you are fine and will at some point come back to InvokeAI 🙈 
Also thx for letting me and others know your opinion while leaving it up to them 👌🏻

File: .editorconfig [hunk 1]
```diff
@@ -1,5 +1,8 @@
+root = true
+
 # All files
 [*]
+max_line_length = 80
```
Comment (keturn) [inline .editorconfig:5]: this should be consistent with the max-line-length in .flake8

**Review by keturn — CHANGES_REQUESTED** [review]

**Comment by keturn** [comment]
isort seems like a good idea.

I object to `black`'s formatting due to this kind of thing:

https://github.com/invoke-ai/InvokeAI/blob/11a29fdc4d3c7dcb9b6bfe56ea22e4122933701f/ldm/invoke/CLI.py#L995-L997

https://github.com/invoke-ai/InvokeAI/blob/11a29fdc4d3c7dcb9b6bfe56ea22e4122933701f/ldm/modules/textual_inversion_manager.py#L411-L413

Those are not good interpretations of any reasonable coding standard.

File: .editorconfig (hunk 1 above)
Comment (mauwii) [reply .editorconfig:5]: added it to python

**Comment by mauwii** [comment]
@keturn I did not run the hook here yet, so the formatting of the files is not black at all (unless the coder used it, but since it is only formatting complete files, not only your changes, this will not be the case).

When I re-format it manually to look like:

```py
def choose_postprocess_name(opt, prefix, seed) -> str:
    match = re.search("postprocess:(\w+)", opt.last_operation)
    if match:
        # will look like "gfpgan", "upscale", "outpaint" or "embiggen"
        modifier = match.group(1)
    else:
        modifier = "postprocessed"
```

black would not touch the file at all 🙈

```sh
black ./ldm/invoke/CLI.py
All done! ✨ 🍰 ✨
1 file left unchanged.
```

**Comment by keturn** [comment]
The examples I quoted were from commits formatted with black. Reproducible with the [playground](https://black.vercel.app/?version=stable&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4AGUAPRdAD2IimZxl1N_WlbvK5V9KEd03r0mK1OKMLFJsqPGy7meyfrF9p2qB1TH5nqEPV0bLtUqPzwL2APm6zbQKJ_orU01eIcPEEB-P1V75I255yr9VVPLZx1to_Era6b8DSXA7JAvAWiytMU0edGDuIfHADVd3pudS63UaDPPWQQLtqNcn8G2GGu750dlB7XTSiicI10w1z0xa_0p1QY2MtgJxAyccCy_oAIDwAHY5WsYPdJoszs9hHS9_1W99CjTw2n4tmGZmfdPtOYZGZHHD689L9_kHE05R_ody1wAJleuQJSAAXjIB8TIlktfINZTS0KzOE1lDwAAvwraKkRmAAgAAZAClQMAADE-KyyxxGf7AgAAAAAEWVo=).

**Comment by mauwii** [comment]
> The examples I quoted were from commits formatted with black. Reproducible with the [playground](https://black.vercel.app/?version=stable&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4AGUAPRdAD2IimZxl1N_WlbvK5V9KEd03r0mK1OKMLFJsqPGy7meyfrF9p2qB1TH5nqEPV0bLtUqPzwL2APm6zbQKJ_orU01eIcPEEB-P1V75I255yr9VVPLZx1to_Era6b8DSXA7JAvAWiytMU0edGDuIfHADVd3pudS63UaDPPWQQLtqNcn8G2GGu750dlB7XTSiicI10w1z0xa_0p1QY2MtgJxAyccCy_oAIDwAHY5WsYPdJoszs9hHS9_1W99CjTw2n4tmGZmfdPtOYZGZHHD689L9_kHE05R_ody1wAJleuQJSAAXjIB8TIlktfINZTS0KzOE1lDwAAvwraKkRmAAgAAZAClQMAADE-KyyxxGf7AgAAAAAEWVo=).

But you saw that the code was already formated like this beforehand, right?

**Comment by keturn** [comment]
Standards are fine. PEP 8 is fine. 

Automatically running every commit through `black` is not a good way to meet those standards.

**Comment by lstein** [comment]
@mauwii Would it be possible to rebase this against `v2.3` ? I'd like to play with this on a slow-moving branch rather than on `main` where a lot of people will be working.

I tried rebasing by pressing the `edit` button above, but the PR ended up containing a bunch of nodes-related files that were carried over from merge commits into the PR (I think).

**Comment by mauwii** [comment]
closed in favor of #2822

**Comment by tildebyte** [comment]
FTR:

> I object to black's formatting due to this kind of thing

1) Absolutely agree
2) Warts like that are usually because the line length is too short (still a black problem, of course; their default is demonstrably ridiculous).  If you force black to use longer lines, things like that are much less common.

**Pull request closed without merging**